pkg/virt-handler/vm.go

```diff
@@ -240,28 +239,66 @@ func (d *VirtualMachineController) updateVMIStatus(vmi *v1.VirtualMachineInstanc
 	oldStatus := vmi.DeepCopy().Status
 
 	if domain != nil {
-
 		// This is needed to be backwards compatible with vmi's which have status interfaces
 		// with the name not being set
-		if len(vmi.Status.Interfaces) == 1 && vmi.Status.Interfaces[0].Name == "" {
+		if len(domain.Spec.Devices.Interfaces) == 0 && len(vmi.Status.Interfaces) == 1 && vmi.Status.Interfaces[0].Name == "" {
 			for _, network := range vmi.Spec.Networks {
 				if network.NetworkSource.Pod != nil {
 					vmi.Status.Interfaces[0].Name = network.Name
 				}
 			}
 		}
 
-		interfacesByName := make(map[string]int)
-		for i, existingInterface := range vmi.Status.Interfaces {
-			interfacesByName[existingInterface.Name] = i
-		}
+		if len(domain.Spec.Devices.Interfaces) > 0 || len(domain.Status.Interfaces) > 0 {
+			newInterfaces := []v1.VirtualMachineInstanceNetworkInterface{}
 
-		for _, domainInterface := range domain.Spec.Devices.Interfaces {
-			if i, exists := interfacesByName[domainInterface.Alias.Name]; exists {
-				vmi.Status.Interfaces[i].MAC = domainInterface.MAC.MAC
-			} else {
-				vmi.Status.Interfaces = append(vmi.Status.Interfaces, v1.VirtualMachineInstanceNetworkInterface{MAC: domainInterface.MAC.MAC, Name: domainInterface.Alias.Name})
+			existingInterfaceStatusByName := map[string]v1.VirtualMachineInstanceNetworkInterface{}
```

Comment:
I am not sure I can completely follow the update logic. Could you explain it?

Reply:
We have 3 sets of info, that we need to consider:
domain.Spec.Devices.Interfaces - the interfaces as defined in spec, always there
domain.Status.Interfaces - interfaces reported by guest agent (can be empty)
vmi.Status.Interfaces - interfaces already reported in vmi (with pod IP)

goal: populate vmi.Status.Interfaces

What we want to have in vmi.Status.Interfaces:
- all interfaces defined in domain.Spec
- any additional interfaces added by user and discovered by guest agent
- on the main pod interface we want to preserver the pod IP (set on vmi.Status by controller)

What we do:
- iterate through all  domain.Spec interfaces
    - for the ones that already exist in vmi.Status, we reuse them (this is needed to keep the pod IP)
    - if not in vmi.Status, we add a new one
    - if interface is present in domain.Status, update the IP info
- iterate though domain.Status interfaces not handled above
  this will add interfaces discovered by guest agent, but not defined in domain.Spec (any interfaces added manually on the vm, like bridges, ...)

Reply:
I'm having hard times keeping the whole function in my brain RAM. Could you please comment every block and say what is going on (or split it to well-named functions?). Make it narrative :)